sycl/doc/extensions/proposed/sycl_ext_oneapi_barrier.asciidoc

```diff
@@ -0,0 +1,430 @@
+
+
+==== Usage examples
+
+As noted above, `barrier` objects must be allocated in memory that is visible
+to all of the work-items using the barrier. There are many ways to satisfy this
+condition. The examples in this section demonstrate some common ways to
+allocate and construct barrier objects for different scopes.
+
+
+===== Work-group scope
+
+Barriers at work-group scope can be allocated in group-local memory. The
+simplest way to allocate and construct a `barrier` object in group-local memory
+is to use the
+link:../supported/sycl_ext_oneapi_local_memory.asciidoc[sycl_ext_oneapi_local_memory]
+extension.
+
+[source,c++]
+----
+using work_group_barrier = syclex::barrier<sycl::memory_scope::work_group>;
+
+q.parallel_for(..., [=](sycl::nd_item it) {
+
+  // Allocate memory for and construct the barrier
+  auto* bar = sycl::ext::oneapi::group_local_memory<work_group_barrier>(it.get_group(), nthreads);
+
+  // Use the barrier
+  bar->arrive_and_wait();
+
+  // The barrier is automatically destructed and the memory is freed upon kernel completion
+
+}).wait();
+----
+
+
+===== Device scope
+
+Barriers at device scope must be allocated in global memory (including various
+kinds of USM). It is recommended that device scope barriers are initialized on
+the device that will use the barrier.
+
+[source,c++]
+----
+using device_barrier = syclex::barrier<sycl::memory_scope::device>;
+
+// Allocate memory for the barrier
+device_barrier* bar = sycl::device_malloc<device_barrier>(1, q);
+
+// Construct the barrier using placement new
+q.single_task([=]() {
+  new (bar) device_barrier(nthreads);
+}).wait();
+
+// Use the barrier
+q.parallel_for(..., [=](sycl::nd_item it) {
+  auto mybar = std::launder(bar);
+  mybar->arrive_and_wait();
+}).wait();
+
+// Destruct the barrier
+q.single_task([=]() {
+  auto mybar = std::launder(bar);
+  mybar->~barrier();
+}).wait();
+
+// Free memory for the barrier
+sycl::free(bar, q);
+----
+
+It is possible to construct and destruct a `barrier` object in the same kernel
+that uses it, rather than using `single_task` as shown above, but this will
+require an additional mechanism of device synchronization such as the root-group
+barrier provided by the
+link:../proposed/sycl_ext_oneapi_root_group.asciidoc[sycl_ext_oneapi_root_group]
+extension.
+
+
+===== System scope
+
+Barriers at system scope must be allocated in host, shared, or system USM. Such
+barriers can be constructed on the host.
+
+[source,c++]
+----
+using system_barrier = syclex::barrier<sycl::memory_scope::system>;
+
+// Allocate memory for the barrier
+system_barrier* bar = sycl::shared_malloc<system_barrier>(1, q);
+
+// Construct the barrier using placement new
+bar = new (bar) system_barrier(nthreads);
+
+// Use the barrier
+q.parallel_for(..., [=](sycl::nd_item it) {
+  bar->arrive_and_wait();
+});
+
+// Destruct the barrier
+bar->~barrier();
+
+// Free memory for the barrier
+sycl::free(bar, q);
+----
+
+
+=== `group_arrive` and `group_wait`
```

Comment:
what would be the difference here between `group_arrive` vs `joint_arrive` vs `arrive_over_group`?
Are we introducing a new concept by using `group` as the prefix here?

Reply:
Good question.  There is precedent for this in SYCL 2020, but it's not obvious.

We use the `joint_` vs `_over_group` distinction for group _algorithms_. These typically (but not always) correspond to `std::` algorithms, and typically accept input values from the calling work-items.  We need to distinguish between the cases where all the work-items are co-operating on a large operation that's uniformly described (i.e. a `joint` algorithm) and the cases where all the work-items are contributing non-uniform input values to an operation (i.e. an `_over_group` algorithm).

There are a few functions in SYCL 2020 where this distinction isn't necessary, because there's only one meaningful interpretation.  SYCL calls these group _functions_, and there are currently only two of them: `group_barrier` and `group_broadcast`.  `group_arrive` and `group_wait` are group functions.  I think it makes sense to describe them that way, and as an added bonus, if they were to make it into a future SYCL standard then they would appear next to `group_barrier` in the spec.

Reply:
I understand what you are saying but I still think the distinction is not clear.
What is the difference between group algorithms and group functions?

As a subsequent question, i see group_arrive and group_wait are now free functions instead of adding them as methods like the WI variant. Is there a reason for that?
Also, what about arrive_and_wait, do we need group_arrive_and_wait too?

Reply:
> I understand what you are saying but I still think the distinction is not clear.
> What is the difference between group algorithms and group functions?

I think it's largely just a judgement call we have to make on a case-by-case basis.  Group algorithms are really a special-case of group functions.  When we introduce new functionality, as long as we have an argument for whether it's described as a function or an algorithm I think that's enough.

> As a subsequent question, i see group_arrive and group_wait are now free functions instead of adding them as methods like the WI variant. Is there a reason for that?

You mean `work_item::barrier`?  That function existed in SYCL 1.2.1, but it was removed in SYCL 2020.

Almost all of the functions that operate on groups are free functions accepting groups as the first parameter.  If I recall correctly, the original inspiration for this was the behavior of the C++17 parallel algorithms, and the behavior of the now defunct executor proposal.  In both cases, the first argument to the algorithms described where/what sort of parallelism should be used by the function.  Similarly, a SYCL `group` object describes how to parallelize the operation.

> Also, what about arrive_and_wait, do we need group_arrive_and_wait too?

I hadn't thought about that.

`group_arrive_and_wait(g, b)` would be equivalent to `group_wait(group_arrive(g, b))`, which is only marginally shorter.  Also, I think the existing `group_barrier(g)` function would cover most of the use-cases for `arrive_and_wait` -- if somebody doesn't need a split barrier, they probably don't need to use `syclex::barrier`.

The only exception I can think of is if we do something similar to NVIDIA, and start tying other operations into the barrier object.  Then `group_barrier` wouldn't be sufficient.

Do you think we need `group_arrive_and_wait` to cover that last case?  It would be easy to write the spec for it, so I'm happy to do it if you think we need it.

Reply:
>>>>> As a subsequent question, i see group_arrive and group_wait are now free functions instead of adding them as methods like the WI variant. Is there a reason for that?

>>> You mean work_item::barrier? That function existed in SYCL 1.2.1, but it was removed in SYCL 2020.

Sorry, I meant like `syclex::barrier::arrive` and `syclex::barrier::wait` so why not adding them as methods to `syclex::barrier`

Reply:
>>>>> Also, what about arrive_and_wait, do we need group_arrive_and_wait too?

>> I hadn't thought about that.

>>Do you think we need group_arrive_and_wait to cover that last case? It would be easy to write the spec for it, so I'm happy to do it if you think we need it.

I am in the process of writing the spec for async_memcpy and in the group variant of this, we would need to pass a barrier object and wait on it. In this case, the whole group waits on that barrier. So I imagined wiritng it like this:
`  group_async_memcpy(g, pdest, psrc, N, bar);`
`  bar->group_arrive_and_wait(g);`

Notice here that `group_arrive` and `group_wait` should take `syclex::barrier` as an argument or be a method of `syclex::barrier`. I am not sure `typename Barrier` is enough.

Reply:
> Sorry, I meant like `syclex::barrier::arrive` and `syclex::barrier::wait` so why not adding them as methods to `syclex::barrier`

Oh, I see.  Two reasons:

1. The interface of `syclex::barrier` is the same as that of `std::barrier` -- `syclex::barrier` has an extra template, but no extra member functions.
3. If we ever reach the point where `std::barrier` can be used inside of device code (which is desirable), we'd be able to update our extension to allow `group_arrive(std::barrier)`. We won't be able to add `group_arrive` to `std::barrier`.

> I am in the process of writing the spec for async_memcpy and in the group variant of this, we would need to pass a barrier object and wait on it.

Makes sense.  I've added `group_arrive_and_wait` in 0ce044f (as a free function).

> Notice here that `group_arrive` and `group_wait` should take `syclex::barrier` as an argument or be a method of `syclex::barrier`. I am not sure `typename Barrier` is enough.

They're currently defined to take `syclex::barrier` as an argument.  The template is required because `syclex::barrier` is itself a templated type, and we need `group_arrive` and `group_wait` to accept barrier objects with any scope and/or completion function.

But you're right that `typename Barrier` is not a sufficient description.  I've added constraints to make it clear that only `syclex::barrier` objects are supported in d513063.